src/allmydata/client.py

```diff
@@ -550,6 +557,14 @@ def tub_creator(handler_overrides=None, **kwargs):
             **kwargs
         )
 
+    # we don't actually use this keypair for anything (yet) as far
+    # as I can see.
+    # my_pubkey = keyutil.parse_pubkey(
+    #     self.get_config_from_file("node.pubkey")
+    # )
```

Comment:
The fact that the pubkey we sign to create grid manager certificates is one of the things that caught me up in thinking about the security properties.  Shouldn't we sign a pubkey that's actually used for something?

Reply:
I don't know what else we'd use besides "the node's identity keypair" .. but also yes I suppose if the private-key isn't used for anything what's the point? Does this tie in to the `grid-introducer` things? Should it demand that announcements from storage-servers are signed by that storage-server?

Reply:
The thing I couldn't figure out (not yet having finished reading the whole PR diff) is what stops an attacker from:

1. Connect to the introducer
2. Receive all the announcements, including certificates
3. Turn around and publish my own announcement for my own storage service and include the stolen certificate

If no part of Tahoe-LAFS uses the private key associated with node.pubkey then I never have to prove that I have the private key associated with the pubkey in the certificate, do I?

But, aha!  Looking deeper I see that the `node.pubkey` / `private/node.privkey` pair *is* used to sign announcements.  If that is true then grid-manager-enabled announcement validation should be a three-step process:

1. The signature on the storage announcement is valid.
2. The grid-manager signature of the grid-manager certificate is valid and uses a key from a trusted grid-manager.
3. The public-key in the grid-manager certificate is the same as the public key associated with the signature from step (1).

(1) is already implemented (in master) in `allmydata.introducer.client.IntroducerClient.got_announcements` (via the call to `unsign_from_foolscap`.  (2) is already implemented (in this PR) in  `allmydata.grid_manager.create_grid_manager_verifier`.  Only (3) is missing.

At least that's a way it could work.  It seems like it might be overly complicated but maybe it's only necessarily complicated... One alternative that suggests itself to me is that the storage server would send its whole announcement to the grid-manager to be signed.  Then it could be a two step process ((1) verify grid-manager's signature (2) verify storage server's signature) but this also increases the complexity to use it ... An announcement is a lot more stuff than is currently being signed.  This makes the CLI more awkward (not least because there's no UI for extracting a storage announcement from a storage server right now).  It also removes flexibility - a storage server would have to go get a new signature any time any detail of its announcement changes.  This would stop it from changing its network address by itself but also would stop it from doing other kinds of reconfiguration, like installing a new storage server plugin (the kind of thing ZKAPAuthorizer is) or changing any configuration details of any existing plugin.  Doesn't seem particularly desirable, overall.